pkg/config/filter.go

```diff
@@ -3,33 +3,70 @@ package config
 import (
 	"fmt"
 	"os"
+	"path/filepath"
 	"strings"
 
 	"github.com/gobwas/glob"
 
 	"github.com/open-policy-agent/opa/bundle"
-	"github.com/open-policy-agent/opa/loader"
 )
 
-func FilterIgnoredPaths(paths, ignore []string, checkFileExists bool) ([]string, error) {
-	policyPaths := paths
+type DirEntryFilter = func(path string, info os.DirEntry, err error) error
 
+func FilterIgnoredPaths(paths, ignore []string, checkFileExists bool) ([]string, error) {
 	if checkFileExists {
-		filteredPaths, err := loader.FilteredPaths(paths, func(_ string, info os.FileInfo, depth int) bool {
-			return !info.IsDir() && !strings.HasSuffix(info.Name(), bundle.RegoExt)
+		filtered := make([]string, 0, len(paths))
+
+		errors := walkPaths(paths, func(path string, info os.DirEntry, err error) error {
+			if info.IsDir() && (info.Name() == ".git" || info.Name() == ".idea") {
+				return filepath.SkipDir
+			}
+			if !info.IsDir() && strings.HasSuffix(path, bundle.RegoExt) {
+				filtered = append(filtered, path)
+			}
+
+			return err
 		})
-		if err != nil {
-			return nil, fmt.Errorf("failed to filter paths: %w", err)
+
+		if len(errors) > 0 {
+			return nil, fmt.Errorf("failed to filter paths: %v", errorsToString(errors))
 		}
 
-		policyPaths = filteredPaths
+		return filterPaths(filtered, ignore)
 	}
 
 	if len(ignore) == 0 {
-		return policyPaths, nil
+		return paths, nil
 	}
 
-	return filterPaths(policyPaths, ignore)
+	return filterPaths(paths, ignore)
+}
+
+// walkPaths is a simplified version of the same function in OPA, but one that
+// avoids stat-ing files in favor of DirEntry introspection. This is faster due to
+// the reduced number of system calls. Since we avoid stat-ing files, we don't get
+// to work with FileInfo objects, but will rather work directly with the files as
+// string paths. Finally, this function differs from OPA in that we never will
+// traverse down into .git or .idea directories, as that's always a waste of time.
+func walkPaths(paths []string, filter DirEntryFilter) []error {
```

Comment:
[nit] It doesn't matter, but golang can join errors now: https://pkg.go.dev/errors#Join

Reply:
Oh, that's interesting! Perhaps we could skip that whole `errorsToString` function entirely then? I'll play around with it.